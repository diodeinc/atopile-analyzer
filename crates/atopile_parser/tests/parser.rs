use atopile_parser::lexer::lex;
use atopile_parser::parser::parse;
use std::fs;

macro_rules! create_parser_test {
    ($name:ident) => {
        #[test]
        fn $name() {
            let input = fs::read_to_string(concat!(
                "tests/resources/corpus/",
                stringify!($name),
                ".ato"
            ))
            .unwrap();

            let normalized_input = input.replace("\r\n", "\n");

            let tokens = lex(&normalized_input).0;
            let output = parse(&tokens);

            for err in output.1.iter() {
                println!("Error: {:?}", err);
                println!("Context tokens:");
                let span = err.span();
                tokens
                    .iter()
                    .enumerate()
                    .filter(|(i, _)| {
                        *i >= span.start.saturating_sub(5) && *i <= span.end.saturating_add(5)
                    })
                    .for_each(|(i, t)| {
                        println!(
                            "{}{:2}: {:?}",
                            if span.into_range().contains(&i) {
                                "-> "
                            } else {
                                "   "
                            },
                            i,
                            t
                        )
                    });
                println!("---");
            }

            insta::assert_debug_snapshot!(output);
        }
    };
}

create_parser_test!(vdivs);
create_parser_test!(resistors);
create_parser_test!(transistors);
create_parser_test!(bma400);
