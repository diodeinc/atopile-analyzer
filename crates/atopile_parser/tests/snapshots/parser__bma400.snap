---
source: crates/atopile_parser/tests/parser.rs
expression: output
---
(
    [
        Spanned(
            Import(
                ImportStmt {
                    from_path: Spanned(
                        "generics/interfaces.ato",
                        1..2,
                    ),
                    imports: [
                        Spanned(
                            "Power",
                            3..4,
                        ),
                        Spanned(
                            "I2C",
                            5..6,
                        ),
                        Spanned(
                            "SPI",
                            7..8,
                        ),
                        Spanned(
                            "GPIO",
                            9..10,
                        ),
                    ],
                },
            ),
            0..10,
        ),
        Spanned(
            Import(
                ImportStmt {
                    from_path: Spanned(
                        "generics/capacitors.ato",
                        12..13,
                    ),
                    imports: [
                        Spanned(
                            "Capacitor",
                            14..15,
                        ),
                    ],
                },
            ),
            11..15,
        ),
        Spanned(
            Import(
                ImportStmt {
                    from_path: Spanned(
                        "_BMA400/BMA400.ato",
                        17..18,
                    ),
                    imports: [
                        Spanned(
                            "BMA400",
                            19..20,
                        ),
                    ],
                },
            ),
            16..20,
        ),
        Spanned(
            Block(
                BlockStmt {
                    kind: Spanned(
                        Module,
                        22..23,
                    ),
                    name: Spanned(
                        "Bma400",
                        23..24,
                    ),
                    parent: None,
                    body: [
                        Spanned(
                            Comment(
                                CommentStmt {
                                    comment: Spanned(
                                        "Ultra-low power triaxial accelerometer driver module based on the BMA400 from Bosch.",
                                        27..28,
                                    ),
                                },
                            ),
                            27..28,
                        ),
                        Spanned(
                            Comment(
                                CommentStmt {
                                    comment: Spanned(
                                        "Features:",
                                        30..31,
                                    ),
                                },
                            ),
                            30..31,
                        ),
                        Spanned(
                            Comment(
                                CommentStmt {
                                    comment: Spanned(
                                        "- Ultra-low power consumption (14.5μA in normal mode, 850nA in low-power mode)",
                                        32..33,
                                    ),
                                },
                            ),
                            32..33,
                        ),
                        Spanned(
                            Comment(
                                CommentStmt {
                                    comment: Spanned(
                                        "- 12-bit digital resolution",
                                        34..35,
                                    ),
                                },
                            ),
                            34..35,
                        ),
                        Spanned(
                            Comment(
                                CommentStmt {
                                    comment: Spanned(
                                        "- Programmable measurement ranges: ±2g, ±4g, ±8g, ±16g",
                                        36..37,
                                    ),
                                },
                            ),
                            36..37,
                        ),
                        Spanned(
                            Comment(
                                CommentStmt {
                                    comment: Spanned(
                                        "- Built-in power management with auto wake-up/sleep capability",
                                        38..39,
                                    ),
                                },
                            ),
                            38..39,
                        ),
                        Spanned(
                            Comment(
                                CommentStmt {
                                    comment: Spanned(
                                        "- Integrated FIFO (1KB)",
                                        40..41,
                                    ),
                                },
                            ),
                            40..41,
                        ),
                        Spanned(
                            Comment(
                                CommentStmt {
                                    comment: Spanned(
                                        "- Smart interrupts including:",
                                        42..43,
                                    ),
                                },
                            ),
                            42..43,
                        ),
                        Spanned(
                            Comment(
                                CommentStmt {
                                    comment: Spanned(
                                        "- Activity/inactivity detection",
                                        44..45,
                                    ),
                                },
                            ),
                            44..45,
                        ),
                        Spanned(
                            Comment(
                                CommentStmt {
                                    comment: Spanned(
                                        "- Step counting with low power consumption (4μA)",
                                        46..47,
                                    ),
                                },
                            ),
                            46..47,
                        ),
                        Spanned(
                            Comment(
                                CommentStmt {
                                    comment: Spanned(
                                        "- Orientation detection",
                                        48..49,
                                    ),
                                },
                            ),
                            48..49,
                        ),
                        Spanned(
                            Comment(
                                CommentStmt {
                                    comment: Spanned(
                                        "- Tap/double-tap sensing",
                                        50..51,
                                    ),
                                },
                            ),
                            50..51,
                        ),
                        Spanned(
                            Comment(
                                CommentStmt {
                                    comment: Spanned(
                                        "- Digital interfaces: I2C or SPI",
                                        52..53,
                                    ),
                                },
                            ),
                            52..53,
                        ),
                        Spanned(
                            Comment(
                                CommentStmt {
                                    comment: Spanned(
                                        "- VDDIO voltage range: 1.2V to 3.6V",
                                        54..55,
                                    ),
                                },
                            ),
                            54..55,
                        ),
                        Spanned(
                            Comment(
                                CommentStmt {
                                    comment: Spanned(
                                        "- VDD voltage range: 1.7V to 3.6V",
                                        56..57,
                                    ),
                                },
                            ),
                            56..57,
                        ),
                        Spanned(
                            Comment(
                                CommentStmt {
                                    comment: Spanned(
                                        "Example Usage:",
                                        59..60,
                                    ),
                                },
                            ),
                            59..60,
                        ),
                        Spanned(
                            Comment(
                                CommentStmt {
                                    comment: Spanned(
                                        "```",
                                        61..62,
                                    ),
                                },
                            ),
                            61..62,
                        ),
                        Spanned(
                            Comment(
                                CommentStmt {
                                    comment: Spanned(
                                        "from 'generics/interfaces.ato' import Power, I2C, GPIO",
                                        63..64,
                                    ),
                                },
                            ),
                            63..64,
                        ),
                        Spanned(
                            Comment(
                                CommentStmt {
                                    comment: Spanned(
                                        "from 'bma400/bma400.ato' import Bma400, Bma400_I2C",
                                        65..66,
                                    ),
                                },
                            ),
                            65..66,
                        ),
                        Spanned(
                            Comment(
                                CommentStmt {
                                    comment: Spanned(
                                        "module UserBoard:",
                                        68..69,
                                    ),
                                },
                            ),
                            68..69,
                        ),
                        Spanned(
                            Comment(
                                CommentStmt {
                                    comment: Spanned(
                                        "'''User board with BMA400 accelerometer'''",
                                        70..71,
                                    ),
                                },
                            ),
                            70..71,
                        ),
                        Spanned(
                            Comment(
                                CommentStmt {
                                    comment: Spanned(
                                        "# Interfaces",
                                        73..74,
                                    ),
                                },
                            ),
                            73..74,
                        ),
                        Spanned(
                            Comment(
                                CommentStmt {
                                    comment: Spanned(
                                        "power_1v8 = new Power  # 1.8V power supply",
                                        75..76,
                                    ),
                                },
                            ),
                            75..76,
                        ),
                        Spanned(
                            Comment(
                                CommentStmt {
                                    comment: Spanned(
                                        "i2c = new I2C          # I2C bus",
                                        77..78,
                                    ),
                                },
                            ),
                            77..78,
                        ),
                        Spanned(
                            Comment(
                                CommentStmt {
                                    comment: Spanned(
                                        "int1 = new GPIO        # Interrupt 1 input",
                                        79..80,
                                    ),
                                },
                            ),
                            79..80,
                        ),
                        Spanned(
                            Comment(
                                CommentStmt {
                                    comment: Spanned(
                                        "# Accelerometer",
                                        82..83,
                                    ),
                                },
                            ),
                            82..83,
                        ),
                        Spanned(
                            Comment(
                                CommentStmt {
                                    comment: Spanned(
                                        "accel = new Bma400",
                                        84..85,
                                    ),
                                },
                            ),
                            84..85,
                        ),
                        Spanned(
                            Comment(
                                CommentStmt {
                                    comment: Spanned(
                                        "accel.ic -> Bma400_I2C  # Use I2C interface",
                                        86..87,
                                    ),
                                },
                            ),
                            86..87,
                        ),
                        Spanned(
                            Comment(
                                CommentStmt {
                                    comment: Spanned(
                                        "# Connections",
                                        89..90,
                                    ),
                                },
                            ),
                            89..90,
                        ),
                        Spanned(
                            Comment(
                                CommentStmt {
                                    comment: Spanned(
                                        "accel.power ~ power_1v8",
                                        91..92,
                                    ),
                                },
                            ),
                            91..92,
                        ),
                        Spanned(
                            Comment(
                                CommentStmt {
                                    comment: Spanned(
                                        "accel.i2c ~ i2c",
                                        93..94,
                                    ),
                                },
                            ),
                            93..94,
                        ),
                        Spanned(
                            Comment(
                                CommentStmt {
                                    comment: Spanned(
                                        "accel.int1 ~ int1",
                                        95..96,
                                    ),
                                },
                            ),
                            95..96,
                        ),
                        Spanned(
                            Comment(
                                CommentStmt {
                                    comment: Spanned(
                                        "```",
                                        97..98,
                                    ),
                                },
                            ),
                            97..98,
                        ),
                        Spanned(
                            Signal(
                                SignalStmt {
                                    name: Spanned(
                                        "gnd",
                                        102..103,
                                    ),
                                },
                            ),
                            101..103,
                        ),
                    ],
                },
            ),
            22..106,
        ),
    ],
    [],
)
